app/src/org/commcare/utils/PlaystoreUtils.kt

```diff
@@ -0,0 +1,24 @@
+package org.commcare.utils
+
+import android.content.Context
+import android.content.Intent
+import android.content.pm.PackageManager
+import android.net.Uri
+import org.commcare.CommCareApplication
+
+object PlaystoreUtils {
+
+    fun isApkInstalled(packageName: String): Boolean {
+        return try {
+            val packageManager = CommCareApplication.instance().packageManager
```

Comment:
What is the failure mode of this message if we don't have **permission** to check for this package?  It seems like that's likely to occur as a source of confusion.

Also, I was thinking that there might be a slightly different way to signal for the installed presence of an app for apps without our "first class" permissions check here. I wonder if we can perform a background callout to a content provider or broadcast mechanism to ask the other app if it is installed. It might be fairly easy to provide that mechanism with the Support Library. Not a "Phase 1" requirement by any means, but worth considering.

Reply:
> What is the failure mode of this message if we don't have permission to check for this package?

It's same as app not being installed on the device i.e. the catched `PackageManager.NameNotFoundException` below. 

>  I wonder if we can perform a background callout to a content provider or broadcast mechanism to ask the other app if it is installed

You mean signalling from External App to CommCare that it's installed ? That sounds feasible for Dimagi maintained apps although it won't work for other generic apps like WhatsApp for example.

Reply:
Huh, unfortunate that Android doesn't differentiate on that for us. It would be much, much better if we were able to flag that an app permission couldn't be checked.

> You mean signaling from External App to CommCare that it's installed ? That sounds feasible for Dimagi maintained apps although it won't work for other generic apps like WhatsApp for example.

Yeah. I'm imagining that part of the required package config model could be, say, a Content Provider endpoint or something similar with a known schema response that could signal that the app is on device.

It wouldn't work with non-CommCare specific apps, but it does give a third party the ability to use the mechanism with equal independence to Dimagi (or us the ability to use the mechanism for a new app we make without or before doing a new mobile release)

Reply:
Got it, that definitely makes sense. I have detailed out this approach [here](https://docs.google.com/document/d/1F_nKK8dSvpEKB3QOqq2A8Rb1AcMKZwpg9_TLK_IrhJ4/edit#heading=h.rup8i9jup2h8).